File: data_structures_and_algorithms/Code_401/class-11/solutions/java/Stack.java
import java.util.*;

class Stack <T> {
    public static void main(String[] args) throws Exception{
        Stack<Integer> stack1 = new Stack<>();
        stack1.push(1);
        stack1.push(3);
        stack1.push(5);
        System.out.println("stack1 = " + stack1);
        System.out.println("stack1.isEmpty() = " + stack1.isEmpty());
        System.out.println("stack1.pop() = " + stack1.pop());
        System.out.println("stack1.peek() = " + stack1.peek());
        System.out.println("stack1.pop() = " + stack1.pop());
        System.out.println("stack1.pop() = " + stack1.pop());
        System.out.println("stack1.isEmpty() = " + stack1.isEmpty());
        System.out.println("stack1 = " + stack1);

    }

    Node<T> top;

    public void push(T value){
        top = new Node<>(value, top);
    }

    public T pop() throws Exception{
        if(top == null) throw new Exception("stack is empty");
        T val = top.value;
        top = top.next;
        return val;
    }

    public T peek() throws Exception{
        if(top == null) throw new Exception("queue is empty");
        return top.value;
    }

    public boolean isEmpty(){
        if(top == null) return true;
        return false;
    }


    public String toString(){
        StringBuilder stringified = new StringBuilder();
        Node<T> current = top;

        while(current != null){
            stringified.append("{" + current.value + "} -> ");
            current = current.next;
        }
        stringified.append("NULL");
        return stringified.toString();
    }

}

File: data_structures_and_algorithms/Code_401/class-10/solutions/python/stacks_and_queues/stack.py
from invalid_operation_error import InvalidOperationError
from node import Node


class Stack:
    def __init__(self):
        self.top = None

    def push(self, value):
        self.top = Node(value, self.top)

    def pop(self):

        if not self.top:
            raise InvalidOperationError("Method not allowed on empty collection")

        popped = self.top

        self.top = self.top.next

        return popped.value

    def is_empty(self):
        return self.top == None

    def peek(self):

        if not self.top:
            raise InvalidOperationError("Method not allowed on empty collection")

        return self.top.value
